tls.go

```diff
@@ -9,23 +9,53 @@ import (
 	"time"
 )
 
-// NewHttpTLSTransport returns a new http.Transport constructed with the TLS config
+// TransportMiddleware represents a way to add an adapter to the existing transport.
+type TransportMiddleware func(*http.Transport) *http.Transport
+
+// TransportConfig holds the configurable values for setting up a http
+// transport.
+type TransportConfig struct {
+	TLSConfig           *tls.Config
+	DisableKeepAlives   bool
+	TLSHandshakeTimeout time.Duration
+	Middlewares         []TransportMiddleware
+}
+
+// NewHTTPTLSTransport returns a new http.Transport constructed with the TLS config
 // and the necessary parameters for Juju.
-func NewHttpTLSTransport(tlsConfig *tls.Config) *http.Transport {
+func NewHTTPTLSTransport(config TransportConfig) *http.Transport {
 	// See https://code.google.com/p/go/issues/detail?id=4677
 	// We need to force the connection to close each time so that we don't
 	// hit the above Go bug.
 	transport := &http.Transport{
-		TLSClientConfig:     tlsConfig,
-		DisableKeepAlives:   true,
-		TLSHandshakeTimeout: 10 * time.Second,
+		TLSClientConfig:     config.TLSConfig,
+		DisableKeepAlives:   config.DisableKeepAlives,
+		TLSHandshakeTimeout: config.TLSHandshakeTimeout,
+	}
+	for _, middlewareFn := range config.Middlewares {
+		transport = middlewareFn(transport)
 	}
-	installHTTPDialShim(transport)
-	registerFileProtocol(transport)
-	installProxyShim(transport)
 	return transport
 }
 
+// DefaultHTTPTransport creates a default transport with HTTP keep alives
+// disabled and proxy middleware enable.
+func DefaultHTTPTransport() *http.Transport {
+	return DefaultHTTPTransportWithMiddlewares([]TransportMiddleware{
+		ProxyMiddleware,
+	})
+}
+
+// DefaultHTTPTransportWithMiddlewares creates a default transport with HTTP
+// keep alives disabled and allows a slice of middlewares to be passed through.
+func DefaultHTTPTransportWithMiddlewares(middlewares []TransportMiddleware) *http.Transport {
+	return NewHTTPTLSTransport(TransportConfig{
+		DisableKeepAlives:   true,
+		TLSHandshakeTimeout: 30 * time.Second,
```

Comment:
You're increasing the timeout value here from the original.  Should we all the HandshakeTimeout, and others to be configurable?  Perhaps allow it as a model config value eventually?

Reply:
OK yeah, agreed!